enhancements/kube-apiserver/stability-point-to-point-network-check.md

```diff
@@ -0,0 +1,327 @@
+	Status ConditionCheckStatus `json:"status"`
+}
+
+type ConditionCheckSpec struct {
+	// SourcePod names the pod from which the condition will be checked
+	SourcePod string `json:"source_pod"`
+	// Endpoint being checked
+	Endpoint string `json:"endpoint"`
+	// Type of check performed on the endpoint
+	Type ConditionCheckType `json:"type"`
+}
+
+type ConditionCheckStatus struct {
+	// Log of check events
+	Log []LogEntry `json:"log"`
+	// Conditions summarize the status of the check
+	Conditions []ConditionCheckCondition `json:"conditions"`
+}
+
+// LogEntry records events
+type LogEntry struct {
+	Time     metav1.Time    `json:"time,omitempty"`
+	Status   LogEntryStatus `json:"status"`
+	Message  string         `json:"message"`
+	Duration time.Duration  `json:"duration"`
+}
+
+type ConditionCheckType string
+
+type LogEntryStatus string
+
+type ConditionCheckConditionType string
+
+type ConditionCheckCondition struct {
+	Type               ConditionCheckConditionType `json:"type"`
+	Status             ConditionStatus             `json:"status"`
+	LastTransitionTime metav1.Time                 `json:"lastTransitionTime,omitempty"`
+	Reason             string                      `json:"reason,omitempty"`
+	Message            string                      `json:"message,omitempty"`
+}
+
+const (
+	Reachable          ConditionCheckConditionType = "Reachable"
+	Pass               LogEntryStatus              = "Pass"
+	Fail               LogEntryStatus              = "Fail"
+	TCPEndpointConnect ConditionCheckType          = "TCPEndpointConnect"
+)
+
+type ConditionStatus string
+
+const (
+	ConditionTrue    ConditionStatus = "True"
+	ConditionFalse   ConditionStatus = "False"
+	ConditionUnknown ConditionStatus = "Unknown"
+)
+
+```
+
+
+#### Point-to-point Network Check Tool
+
+Runs in a container in the interested pod. Executes check and updates `ConditionCheck` resources for the pod.
+
+Usage: 
+
+```
+p2pnc check tcp-endpoint [list of endpoints]
+
+``` 
+
+Example:
+
+```
+p2pnc check tcp-endpoint \
+    10.0.100.1:2379 \
+    10.0.200.3:2379 \
+    10.0.300.5:2379
+```
+
+Running the above example command would create and then update a `ConditionC`
+
+```yaml
+Kind: ConditionCheck
+Version: network.openshift.io/v1alpha1
+Metadata:
+    Name: kas-10-0-137-239-r7-to-etcd-10-0-148-99
+    Namespace: openshift-kube-apiserver
+Status:
+    Conditions:
+      - Type: Reachable
+        Status: False
+        Message: "Failed connect to 10.0.134.23:2379; No route to host"
+    Log:
+      - Time: '2020-04-20T14:11:18Z'
+        Status: Fail
+        Message: "Failed connect to 10.0.134.23:2379; No route to host"
+        Duration: "208ms"
+      - Time: '2020-04-20T13:11:18Z'
+        Status: Pass
+        Message: "Connected to 10.0.134.23:2379"
+        Duration: "200ms"
+      - Time: '2020-04-20T12:11:18Z'
+        Status: Pass
+        Message: "Connected to 10.0.134.23:2379"
+        Duration: "200ms"
+```
+
+### Risks and Mitigations
```

Comment:
i am curious if a connection failure exists for any of the scenarios outlined in the goal, are we able to actually record the result (e.g. example, kas to etcd).  i am also wondering about the impact of using this feature if it was used by everyone across the cluster at scale.  was anything explored to record the output to a log of some kind instead.  was that evaluated and set aside for my own awareness so i can catch up?

Reply:
who garbage collects this resource?

Reply:
We will provide a pruning controller that can be plugged into interested operators.

Reply:
> i am curious if a connection failure exists for any of the scenarios outlined in the goal, are we able to actually record the result (e.g. example, kas to etcd). i am also wondering about the impact of using this feature if it was used by everyone across the cluster at scale. was anything explored to record the output to a log of some kind instead. was that evaluated and set aside for my own awareness so i can catch up?

Interestingly, the answer is "usually".  Done properly, this will track a current state and hold historical outages.  When it is able to write again, it will be able to post all of its updates, so we can see historical issues.  Most of our issues appear to be intermittent.  Also, many of them involve connections not related to etcd access.

In addition, a separate deadman controller can be written to indicate a condition when logentries stop getting added.

@sanchezl worth recording this.

Reply:
Updated the description of the binary.